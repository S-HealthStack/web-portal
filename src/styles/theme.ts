import { css, DefaultTheme } from 'styled-components';
import { px } from './utils';

const colors = {
  // gray gradations
  black: '#000000',
  black40: 'rgba(0, 0, 0, 0.4)',
  black08: 'rgba(0, 0, 0, 0.08)',

  // light mode
  surface: '#FFFFFF',
  background: '#F7F8FA',
  onPrimary: '#FFFFFF',
  backgroundOnPrimary: '#FFFFFF',
  backgroundSurface: '#FFFFFF',
  backgroundOnSurface: '#2D2D2D',
  primary: '#4475E3',
  primaryLight: '#ECF1FC',
  primaryHovered: '#2F65E0',
  primary05: '#F6F8FE',
  primary10: '#ECF1FC',
  primary30: '#C7D6F7',
  primaryDisabled: '#B3C6F1',
  primaryLightPressed: '#E3EAFB',
  primaryLightFocused: '#C7D6F7',
  textPrimary: '#474747',
  primaryWhite: '#FFFFFF',
  disabled: '#DBDBDB',
  disabled20: 'rgba(219, 219, 219, 0.2)',
  onDisabled: '#A1A1A1',

  textPrimaryBlue: '#4475E3',
  textPrimaryDark: '#474747',
  textSecondaryGray: '#808080',
  textDisabled: '#838383',
  onSurface: '#2D2D2D',
  secondarySkyblue: '#00B0D7', // replace
  backgoundLight: '#F0F0F0',
  primaryBlueHovered: '#2F65E0',
  primaryBluePressed: '#2057D5',

  statusSuccess: '#2DC008',
  statusSuccess10: '#DCF2EA',
  statusSuccessText: '#00825D',

  statusError: '#FF3333',
  statusError10: '#FFEBEB',
  statusErrorText: '#D14343',

  statusWarning: '#F5BF00',
  statusWarning10: '#FFF7D0',
  statusWarningText: '#A36400',

  secondaryViolet: '#944ED7',
  secondaryVioletHovered: '#6A3799',
  secondaryViolet10: '#F4EDFB',

  secondaryTangerine: '#F39C18',
  secondaryTangerineHovered: '#CC8314',
  secondaryTangerine10: '#FEF5E8',

  secondarySkyBlue: '#00B0D7',
  secondarySkyBlueHovered: '#0092B3',
  secondarySkyBlue10: '#E5F7FB',

  secondaryGreen: '#1AD598',
  secondaryGreenHovered: '#16B37F',
  secondaryGreen10: '#E8FBF5',

  secondaryRed: '#FA5F4F',
  secondaryRedHovered: '#CC4E40',
  secondaryRed10: '#FFEFED',
};

export type SpecColorType = keyof typeof colors;

const weightToCssValue = {
  regular: 400,
  upperregular: 450,
  medium: 500,
  semibold: 600,
  bold: 700,
};

const fontSpec = (
  size: number,
  lineHeight: number | string,
  weight: keyof typeof weightToCssValue,
  fontFamily: string
) => css`
  font-family: ${fontFamily}, sans-serif;
  font-size: ${px(size)};
  line-height: ${typeof lineHeight === 'string' ? lineHeight : px(lineHeight)};
  font-weight: ${weightToCssValue[weight]};
  font-style: normal;
`;

const typography = {
  // Inter typeface
  labelRegular: fontSpec(10, 15, 'regular', 'Inter'),
  labelSemibold: fontSpec(10, 15, 'semibold', 'Inter'),

  bodyXSmallRegular: fontSpec(12, 18, 'regular', 'Inter'),
  bodyXSmallSemibold: fontSpec(12, 18, 'semibold', 'Inter'),
  bodySmallRegular: fontSpec(14, 21, 'regular', 'Inter'),
  bodySmallSemibold: fontSpec(14, 21, 'semibold', 'Inter'),
  bodyMediumRegular: fontSpec(16, 24, 'regular', 'Inter'),
  bodyMediumSemibold: fontSpec(16, 24, 'semibold', 'Inter'),
  bodyLargeRegular: fontSpec(18, 26, 'regular', 'Inter'),
  bodyLargeSemibold: fontSpec(18, 26, 'semibold', 'Inter'),

  headingXSmall: fontSpec(16, 21, 'semibold', 'Inter'),
  headingSmall: fontSpec(18, 23, 'semibold', 'Inter'),
  headingXMedium: fontSpec(18, 27, 'semibold', 'Inter'),
  headingXMediumRegular: fontSpec(18, 27, 'regular', 'Inter'),
  headingMedium: fontSpec(24, 32.2, 'semibold', 'Inter'),
  headingLargeSemibold: fontSpec(32, 42, 'semibold', 'Inter'),
  headingXLargeSemibold: fontSpec(48, 58, 'semibold', 'Inter'),
  headingXXLargeSemibold: fontSpec(60, 72, 'semibold', 'Inter'),

  // Fira Code typeface
  query14: fontSpec(14, '130%', 'upperregular', 'Fira Code'),
  query12: fontSpec(12, '150%', 'upperregular', 'Fira Code'),
};

const boxShadow = {
  card: `0 0 ${px(2)} ${px(0)} rgba(0,0,0,0.15)`,
  avatar: `0 ${px(8)} ${px(10)} ${px(2)} 0 rgba(0,0,0,0.12)`,
  studyAvatar: `0 ${px(15)} ${px(20)} rgba(180,180,180,0.3)`,
  previewScreen: `0 ${px(2)} ${px(4)} rgba(71,71,71,0.25)`,
};

const animation = {
  defaultTiming: 'cubic-bezier(0.45, 0.05, 0, 1)',
  rippleTiming: 'cubic-bezier(0.4, 0, 0.2, 1)',
};

declare module 'styled-components' {
  export interface DefaultTheme {
    colors: typeof colors;
    typography: typeof typography;
    boxShadow: typeof boxShadow;
    animation: typeof animation;
  }
}

const theme: DefaultTheme = {
  colors,
  typography,
  boxShadow,
  animation,
};

export default theme;
