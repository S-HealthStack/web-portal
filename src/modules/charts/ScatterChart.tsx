import React, { useRef, useMemo, useState, useCallback } from 'react';
import { useTheme } from 'styled-components';
import * as d3 from 'd3';
import _isEqual from 'lodash/isEqual';

import { SpecColorType } from 'src/styles/theme';
import Tooltip from 'src/common/components/Tooltip';
import {
  getFocusHeight,
  getFocusWidth,
  getDefaultXRange,
  getDefaultYRange,
  getTooltipPosition,
  MARGIN_FOCUS,
  Container,
  TooltipContent,
  DOTS_CONTAINER_CLASS_NAME,
  TooltipProps,
  TooltipNameBlock,
  TooltipColorPoint,
  TooltipDataBlock,
} from './common-helpers';
import { Line, Dot, DotDataItem, XAxis, YAxis, Zoom } from './common-components';

export interface DotItem {
  name: string;
  age: number;
  value: number;
  lastSync: number;
  color: SpecColorType;
}

export interface LineItem {
  name: string;
  age: number;
  value: number;
  color: SpecColorType;
}

type Props = {
  dots: DotItem[];
  lines: LineItem[];
  width: number;
  height: number;
  showTrendLine?: boolean;
  hiddenDataLines: string[];
};

const BASIC_DOT_SIZE = 8;

const getDotId = (index: number) => `dot-${index}`;

const ScatterChart = ({ dots, lines, width, height, showTrendLine, hiddenDataLines }: Props) => {
  const svgRef = useRef<SVGSVGElement>(null);

  const containerRef = useRef<HTMLDivElement>(null);
  const allowTooltipUpdateRef = useRef(false);
  const lastTooltipProps = useRef<TooltipProps>(null);
  const lastEnteredDotId = useRef<string>('');
  const theme = useTheme();

  const [tooltipProps, setTooltipProps] = useState<TooltipProps>(null);
  const [enteredDotId, setEnteredDotId] = useState<string>('');
  const [tooltipDisabled, setTooltipDisabled] = useState(false);

  const linesData = useMemo(() => {
    const groupedData = d3.group(lines, (d) => d.name);
    const groupNames = d3.map(groupedData, (d) => d[0]);
    const sortedLinesData = d3.map(groupedData, (d) =>
      d[1].sort((item1, item2) => item1.age - item2.age)
    );

    const colorScale = d3.scaleOrdinal(
      groupNames,
      d3.map(groupedData, (d) => theme.colors[d[1][0].color])
    );

    return { sortedLinesData, colorScale };
  }, [lines, theme.colors]);

  const xStartDomain = useMemo(() => {
    const minAge = d3.min(dots, (d) => d.age) || 0;
    const maxAge = d3.max(dots, (d) => d.age) || 0;

    return [Math.floor(minAge / 10) * 10, Math.ceil(maxAge / 10) * 10];
  }, [dots]);

  const yStartDomain = useMemo(() => {
    const minValue = d3.min(dots, (d) => +d.value) || 0;
    const maxValue = d3.max(dots, (d) => +d.value) || 0;

    return [Math.floor(minValue / 10) * 10, Math.ceil(maxValue / 10) * 10];
  }, [dots]);

  const [selectedXDomain, setSelectedXDomain] = useState(xStartDomain);
  const [selectedYDomain, setSelectedYDomain] = useState(yStartDomain);

  const xScale = useMemo(
    () => d3.scaleLinear().domain(xStartDomain).range(getDefaultXRange(width)),
    [xStartDomain, width]
  );

  const xScaleSelected = useMemo(
    () => d3.scaleLinear().domain(selectedXDomain).range(getDefaultXRange(width)),
    [selectedXDomain, width]
  );

  const yScale = useMemo(
    () => d3.scaleLinear().domain(yStartDomain).range(getDefaultYRange(height)),
    [yStartDomain, height]
  );

  const yScaleSelected = useMemo(
    () => d3.scaleLinear().domain(selectedYDomain).range(getDefaultYRange(height)),
    [selectedYDomain, height]
  );

  const calculateTooltipPoint = (
    svgRect: DOMRect,
    dotRect: DOMRect,
    position: ReturnType<typeof getTooltipPosition>
  ): [number, number] => {
    switch (position) {
      case 'b':
        return [
          dotRect.x - svgRect.left + dotRect.width / 2,
          dotRect.y - svgRect.top + dotRect.height,
        ];

      case 't':
        return [dotRect.x - svgRect.left + dotRect.width / 2, dotRect.y - svgRect.top];

      case 'l':
        return [dotRect.x - svgRect.left, dotRect.y - svgRect.top + dotRect.height / 2];

      case 'r':
        return [
          dotRect.x - svgRect.left + dotRect.width,
          dotRect.y - svgRect.top + dotRect.height / 2,
        ];

      default:
        return [0, 0];
    }
  };

  const updateXDomain = useCallback(
    (newDomain: number[]) => {
      if (!_isEqual(newDomain, selectedXDomain)) {
        setSelectedXDomain(newDomain);
        if (
          allowTooltipUpdateRef.current &&
          lastEnteredDotId.current &&
          lastTooltipProps.current &&
          svgRef.current
        ) {
          const svgRect = svgRef.current.getBoundingClientRect();
          const dotNode = d3
            .select(svgRef.current)
            .select<Element>(`#${lastEnteredDotId.current}`)
            .node()
            ?.getBoundingClientRect();

          if (dotNode) {
            setTooltipProps({
              ...lastTooltipProps.current,
              point: calculateTooltipPoint(
                svgRect,
                dotNode,
                lastTooltipProps.current.position as ReturnType<typeof getTooltipPosition>
              ),
            });
          }
        }
      }
    },
    [selectedXDomain]
  );

  const updateYDomain = useCallback(
    (newDomain: number[]) => {
      if (!_isEqual(newDomain, selectedYDomain)) {
        setSelectedYDomain(newDomain);
      }
    },
    [selectedYDomain]
  );

  const onDotMouseEnter = useCallback(
    (event: React.MouseEvent<SVGPathElement, MouseEvent>, d: DotDataItem) => {
      if (!svgRef.current) {
        return;
      }

      allowTooltipUpdateRef.current = true;

      const svgRect = svgRef.current.getBoundingClientRect();
      const dotNode = d3
        .select(svgRef.current)
        .select<Element>(`#${d.id}`)
        .node()
        ?.getBoundingClientRect();
      const color = linesData.colorScale(d.name);

      const tooltipContent = (
        <TooltipContent>
          <TooltipNameBlock>
            <TooltipColorPoint color={color} />
            <p>{`${d.name}:` /* TODO: replace by pacticipant id */}</p>
          </TooltipNameBlock>
          <TooltipDataBlock>
            <p>{Math.round(d.y)} bpm</p>
          </TooltipDataBlock>
        </TooltipContent>
      );

      const position = getTooltipPosition(event.pageX, event.pageY, svgRect);

      const tProps = {
        content: tooltipContent,
        point: dotNode
          ? calculateTooltipPoint(svgRect, dotNode, position)
          : [event.pageX - svgRect.left, event.pageY - svgRect.top],
        position,
      } as TooltipProps;

      setTooltipProps(tProps);
      lastTooltipProps.current = tProps;

      if (d.id) {
        setEnteredDotId(d.id);
        lastEnteredDotId.current = d.id;
      }
    },
    [linesData]
  );

  const resetSelectedDot = useCallback(() => {
    setTooltipProps(null);
    setEnteredDotId('');
    lastEnteredDotId.current = '';
    lastTooltipProps.current = null;
  }, []);

  const omDotMouseLeave = useCallback(() => {
    allowTooltipUpdateRef.current = false;
    resetSelectedDot();
  }, [resetSelectedDot]);

  const getDotOpacity = useCallback(
    (d: DotItem, dotId: string) => {
      if (hiddenDataLines?.includes(d.name)) {
        return 0;
      }

      if (!tooltipProps || enteredDotId === dotId || tooltipDisabled) {
        return 1;
      }

      return 0.5;
    },
    [enteredDotId, hiddenDataLines, tooltipDisabled, tooltipProps]
  );

  const xAxis = useMemo(
    () => (
      <XAxis
        key="x-axis"
        xScale={xScaleSelected}
        tickFormatFn={(d: d3.NumberValue) => `${d} yrs`}
        yTickOffset={20}
        yOffset={height - MARGIN_FOCUS.bottom}
        tickSize={getFocusHeight(height)}
        ticks={6}
        removeDomain
      />
    ),
    [height, xScaleSelected]
  );

  const yAxis = useMemo(
    () => (
      <YAxis
        key="y-axis"
        yScale={yScaleSelected}
        tickSize={getFocusWidth(width)}
        xOffset={MARGIN_FOCUS.left}
        xTickOffset={-20}
        ticks={5}
        removeDomain
      />
    ),
    [width, yScaleSelected]
  );

  const lineItems = useMemo(
    () =>
      linesData.sortedLinesData.map((lineItem) => {
        const id = `line-${lineItem[0].name}`;
        return (
          <Line
            key={id}
            id={id}
            xScale={xScaleSelected}
            yScale={yScaleSelected}
            data={lineItem.map((d) => ({ x: d.age, y: d.value }))}
            color={linesData.colorScale(lineItem[0].name)}
            width={4}
            strokeOpacity={
              // eslint-disable-next-line no-nested-ternary
              !hiddenDataLines?.includes(lineItem[0].name) && showTrendLine
                ? tooltipProps
                  ? 0.5
                  : 1
                : 0
            }
          />
        );
      }),
    [hiddenDataLines, linesData, showTrendLine, tooltipProps, xScaleSelected, yScaleSelected]
  );

  const k = useMemo(() => {
    const xDomain = xScale.domain();
    const xSelectedDomain = xScaleSelected.domain();
    const yDomain = yScale.domain();
    const ySelectedDomain = yScaleSelected.domain();
    const xSelectedDomainSize = xSelectedDomain[1] - xSelectedDomain[0];
    const ySelectedDomainSize = ySelectedDomain[0] - ySelectedDomain[1];

    const kX = xSelectedDomainSize ? (xDomain[1] - xDomain[0]) / xSelectedDomainSize : 1;
    const kY = ySelectedDomainSize ? (yDomain[0] - yDomain[1]) / ySelectedDomainSize : 1;

    return Math.max(kX, kY);
  }, [xScale, xScaleSelected, yScale, yScaleSelected]);

  const dotItems = useMemo(
    () => (
      <g className={DOTS_CONTAINER_CLASS_NAME} key="dots">
        {dots.map((d, index) => (
          <Dot
            key={getDotId(index)}
            id={getDotId(index)}
            xScale={xScaleSelected}
            yScale={yScaleSelected}
            size={BASIC_DOT_SIZE * k}
            color={theme.colors[d.color]}
            fillOpacity={getDotOpacity(d, `dot-${index}`)}
            data={{ x: d.age, y: d.value, lastSync: d.lastSync, name: d.name }}
            hoverDisabled={tooltipDisabled}
            onMouseEnter={onDotMouseEnter}
            onMouseLeave={omDotMouseLeave}
          />
        ))}
      </g>
    ),
    [
      dots,
      getDotOpacity,
      k,
      omDotMouseLeave,
      onDotMouseEnter,
      theme.colors,
      tooltipDisabled,
      xScaleSelected,
      yScaleSelected,
    ]
  );

  if (!width || !height || !dots.length || !lines.length) {
    return null;
  }

  return (
    <Container width={width} height={height} ref={containerRef}>
      <Zoom
        svgRef={svgRef}
        width={width}
        height={height}
        showZoomControls={false}
        x={xScaleSelected}
        y={yScaleSelected}
        xContext={xScale}
        yContext={yScale}
        transformK={k}
        childrenBefore={[xAxis, yAxis, lineItems]}
        childrenAfter={[dotItems]}
        updateXDomain={updateXDomain}
        updateYDomain={updateYDomain}
        setChartTooltipDisabled={setTooltipDisabled}
      />
      {tooltipProps && !tooltipDisabled && (
        <Tooltip
          static
          show
          arrow
          content={tooltipProps?.content}
          point={tooltipProps?.point}
          position={tooltipProps?.position}
        />
      )}
    </Container>
  );
};

export default ScatterChart;
