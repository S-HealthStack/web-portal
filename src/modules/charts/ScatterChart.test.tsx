import React from 'react';
import '@testing-library/jest-dom';
import 'jest-styled-components';
import { act, render, fireEvent, findByTestId, waitFor } from '@testing-library/react';
import { ThemeProvider } from 'styled-components/';
import userEvent from '@testing-library/user-event';
import theme from 'src/styles/theme';
import ScatterChart from 'src/modules/charts/ScatterChart';

describe('ScatterChart', () => {
  it('should render', async () => {
    const { baseElement, getByTestId } = render(
      <ThemeProvider theme={theme}>
        <ScatterChart
          width={1000}
          height={1000}
          dots={[
            {
              name: 'test',
              age: 30,
              value: 150,
              lastSync: 0,
              color: 'primary',
            },
          ]}
          hiddenDataLines={[]}
          lines={[
            {
              name: 'test',
              age: 30,
              value: 180,
              color: 'primary',
            },
            {
              name: 'test',
              age: 40,
              value: 140,
              color: 'primary',
            },
          ]}
        />
      </ThemeProvider>
    );
    expect(baseElement).toMatchSnapshot();

    const dot = await findByTestId(baseElement, 'dot-0');
    expect(dot).toBeInTheDocument();

    await act(async () => {
      await userEvent.hover(dot);
    });

    await act(async () => {
      await userEvent.unhover(dot);
    });

    const zoomInButton = getByTestId('zoom-in-button');
    const zoom = baseElement.querySelector('.zoom') as HTMLElement;

    expect(zoom).toBeInTheDocument();
    expect(zoomInButton).toBeInTheDocument();

    await act(async () => {
      await userEvent.click(zoomInButton);
    });

    await waitFor(async () => expect(zoom).toHaveStyle('pointer-events: all'));

    expect(zoomInButton.querySelector('svg')).toHaveStyle('fill: #4475E3');

    fireEvent(
      zoom,
      new MouseEvent('mousedown', {
        bubbles: true,
        cancelable: true,
        clientX: 610,
        clientY: 180,
      })
    );
    fireEvent(
      zoom,
      new MouseEvent('mousemove', {
        bubbles: true,
        cancelable: true,
        clientX: 640,
        clientY: 200,
      })
    );
    fireEvent(
      zoom,
      new MouseEvent('mouseup', {
        bubbles: true,
        cancelable: true,
      })
    );

    await waitFor(async () =>
      expect(zoomInButton.querySelector('svg')).toHaveStyle('fill: #474747')
    );

    const zoomOutButton = getByTestId('zoom-out-button');

    await act(async () => {
      await userEvent.click(zoomOutButton);
    });

    await waitFor(async () =>
      expect(zoomOutButton.querySelector('svg')).toHaveStyle('fill: #DBDBDB')
    );
  });

  it('[NEGATIVE] should render with empty data', () => {
    const { baseElement } = render(
      <ThemeProvider theme={theme}>
        <ScatterChart width={1000} height={1000} dots={[]} hiddenDataLines={[]} lines={[]} />
      </ThemeProvider>
    );

    expect(baseElement).toBeDefined();
  });
});
