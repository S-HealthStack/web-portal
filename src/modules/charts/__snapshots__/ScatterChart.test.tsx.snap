// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScatterChart should render 1`] = `
.c9 {
  display: inline-block;
  position: relative;
}

.c3 .xAxis .tick line {
  opacity: 0;
}

.c3 .xAxis .tick:first-child text {
  text-anchor: start;
}

.c3 .xAxis .tick:last-child text {
  text-anchor: end;
}

.c3 .yAxis .tick line {
  stroke: #F0F0F0;
}

.c3 .context {
  opacity: 0;
  -webkit-transition: opacity 0.3s cubic-bezier(0.45,0.05,0,1);
  transition: opacity 0.3s cubic-bezier(0.45,0.05,0,1);
}

.c3 .context .brushX .overlay {
  fill: #4475E3;
  fill-opacity: 0.1;
  pointer-events: none;
  height: 0.25rem;
}

.c3 .context .brushX .selection {
  fill: #4475E3;
  fill-opacity: 1;
  height: 1.375rem;
  fill: url(#horizontalGradient);
}

.c3 .context .brushX .handle {
  stroke-width: 0.0625rem;
  stroke: #4475E3;
  stroke-opacity: 0.3;
  fill: #FFFFFF;
  height: 1.375rem;
  width: 0.375rem;
}

.c3 .context .brushY .overlay {
  fill: #4475E3;
  fill-opacity: 0.1;
  pointer-events: none;
  width: 0.25rem;
}

.c3 .context .brushY .selection {
  fill: #4475E3;
  fill-opacity: 1;
  width: 1.375rem;
  fill: url(#verticalGradient);
}

.c3 .context .brushY .handle {
  stroke-width: 0.0625rem;
  stroke: #4475E3;
  stroke-opacity: 0.3;
  fill: #FFFFFF;
  height: 0.375rem;
  width: 1.375rem;
}

.c3 .focus {
  cursor: move;
}

.c3 .line {
  -webkit-clip-path: url(#clip);
  clip-path: url(#clip);
  -webkit-transition: stroke-opacity 0.3s cubic-bezier(0.45,0.05,0,1);
  transition: stroke-opacity 0.3s cubic-bezier(0.45,0.05,0,1);
}

.c3 .area {
  -webkit-clip-path: url(#clip);
  clip-path: url(#clip);
  -webkit-transition: fill-opacity 0.3s cubic-bezier(0.45,0.05,0,1);
  transition: fill-opacity 0.3s cubic-bezier(0.45,0.05,0,1);
}

.c3 .dots {
  -webkit-clip-path: url(#clip);
  clip-path: url(#clip);
}

.c3 .dots .dot {
  -webkit-transition: fill-opacity 0.3s cubic-bezier(0.45,0.05,0,1);
  transition: fill-opacity 0.3s cubic-bezier(0.45,0.05,0,1);
}

.c5 > .tick > text {
  font-family: Inter,sans-serif;
  font-size: 0.625rem;
  line-height: 0.9375rem;
  font-weight: 400;
  font-style: normal;
}

.c6 > .tick > text {
  font-family: Inter,sans-serif;
  font-size: 0.625rem;
  line-height: 0.9375rem;
  font-weight: 400;
  font-style: normal;
}

.c1 {
  background: none;
  border: none;
  padding: 0;
  height: 1.5rem;
  width: 1.5rem;
}

.c1 svg {
  fill: #808080;
}

.c1:active,
.c1:enabled:hover {
  cursor: pointer;
}

.c1:active svg,
.c1:enabled:hover svg {
  fill: #808080;
}

.c1:disabled svg {
  fill: #DBDBDB;
}

.c7 {
  position: absolute;
  top: -4.625rem;
  right: 0;
  display: inline-grid;
  grid-template-columns: 1.5rem 1.5rem;
  grid-gap: 0.5rem;
  z-index: 100;
  height: 1.5rem;
}

.c8 {
  cursor: default;
}

.c8 svg {
  fill: #DBDBDB;
}

.c10 {
  cursor: pointer;
}

.c10 svg {
  fill: #474747;
}

.c10 svg:active {
  fill: #4475E3;
}

.c0 {
  position: relative;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 1.5rem;
  height: 1.5rem;
  position: absolute;
  right: 0.25rem;
  top: 0.25rem;
  border-radius: 0.25rem;
  opacity: 0;
  pointer-events: none;
  -webkit-transition: opacity 0.3s cubic-bezier(0.45,0.05,0,1);
  transition: opacity 0.3s cubic-bezier(0.45,0.05,0,1);
  cursor: default;
}

.c4 .zoomRect {
  stroke: #000000;
  fill: transparent;
}

.c4 .focus {
  fill: none;
}

.c4 .focus .zoom {
  pointer-events: none;
}

<body>
  <div>
    <div
      class=""
      height="1000"
      style="width: 62.5rem; height: 62.5rem; position: relative;"
      width="1000"
    >
      <div
        class="c0"
      >
        <button
          aria-label="Zoom Out"
          class="c1 c2"
          color="textSecondaryGray"
        >
          <svg
            data-testid="icon"
          />
        </button>
        <svg
          class="c3 c4"
          viewBox="0, 0, 1000, 1000"
        >
          <g
            class="c5 xAxis undefined"
            fill="none"
            font-family="sans-serif"
            font-size="10"
            text-anchor="middle"
            transform="translate(0, 976)"
          >
            <g
              class="tick"
              opacity="1"
              transform="translate(505.5,0)"
            >
              <line
                stroke="currentColor"
                y2="-971"
              />
              <text
                dy="0em"
                fill="currentColor"
                y="20"
              >
                30 yrs
              </text>
            </g>
          </g>
          <g
            class="c6 yAxis"
            fill="none"
            font-family="sans-serif"
            font-size="10"
            text-anchor="start"
            transform="translate(22,0)"
          >
            <g
              class="tick"
              opacity="1"
              transform="translate(0,491)"
            >
              <line
                stroke="currentColor"
                x2="966"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-20"
              >
                150
              </text>
            </g>
          </g>
          <path
            class="line"
            d="M505,490.5L505,490.5"
            data-testid="line-test"
            fill="none"
            id="line-test"
            stroke="#4475E3"
            stroke-opacity="0"
            stroke-width="2"
          />
          <g
            class="focus"
            style="cursor: grab;"
          >
            <rect
              fill="transparent"
              height="971"
              transform="translate(22, 5)"
              width="966"
            />
            <rect
              class="zoom"
              height="971"
              transform="translate(22, 5)"
              width="966"
            />
            <g
              class="dots"
            >
              <circle
                class="dot"
                cursor="pointer"
                data-testid="dot-0"
                fill="#4475E3"
                fill-opacity="1"
                id="dot-0"
                pointer-events="all"
                r="4"
                transform="translate(505,490.5)"
              />
            </g>
          </g>
          <defs>
            <lineargradient
              id="horizontalGradient"
              x1="0"
              x2="0"
              y1="0"
              y2="1"
            >
              <stop
                offset="0%"
                stop-color="transparent"
              />
              <stop
                offset="41%"
                stop-color="transparent"
              />
              <stop
                offset="41%"
                stop-color="#4475E3"
              />
              <stop
                offset="50%"
                stop-color="#4475E3"
              />
              <stop
                offset="59%"
                stop-color="#4475E3"
              />
              <stop
                offset="59%"
                stop-color="transparent"
              />
              <stop
                offset="100%"
                stop-color="transparent"
              />
            </lineargradient>
            <lineargradient
              id="verticalGradient"
            >
              <stop
                offset="0%"
                stop-color="transparent"
              />
              <stop
                offset="41%"
                stop-color="transparent"
              />
              <stop
                offset="41%"
                stop-color="#4475E3"
              />
              <stop
                offset="50%"
                stop-color="#4475E3"
              />
              <stop
                offset="59%"
                stop-color="#4475E3"
              />
              <stop
                offset="59%"
                stop-color="transparent"
              />
              <stop
                offset="100%"
                stop-color="transparent"
              />
            </lineargradient>
            <clippath
              id="clip"
              transform="translate(22, 5)"
            >
              <rect
                height="971"
                width="966"
              />
            </clippath>
          </defs>
          <g
            class="context"
            display="none"
          >
            <g
              class="brushX"
              fill="none"
              pointer-events="all"
            >
              <rect
                class="overlay"
                cursor="crosshair"
                height="4"
                pointer-events="all"
                transform="translate(0, -1)"
                width="966"
                x="22"
                y="0.5"
              />
              <rect
                class="selection"
                cursor="move"
                fill="#777"
                fill-opacity="0.3"
                height="4"
                shape-rendering="crispEdges"
                stroke="#fff"
                style=""
                transform="translate(0, -10)"
                width="966"
                x="22"
                y="0.5"
              />
              <rect
                class="handle handle--w"
                cursor="ew-resize"
                height="10"
                rx="1"
                style=""
                transform="translate(0, -7)"
                width="6"
                x="19"
                y="-2.5"
              />
              <rect
                class="handle handle--e"
                cursor="ew-resize"
                height="10"
                rx="1"
                style=""
                transform="translate(0, -7)"
                width="6"
                x="985"
                y="-2.5"
              />
            </g>
            <g
              class="brushY"
              fill="none"
              pointer-events="all"
            >
              <rect
                class="overlay"
                cursor="crosshair"
                height="971"
                pointer-events="all"
                transform="translate(2)"
                width="4"
                x="992"
                y="24"
              />
              <rect
                class="selection"
                cursor="move"
                fill="#777"
                fill-opacity="0.3"
                height="971"
                shape-rendering="crispEdges"
                stroke="#fff"
                style=""
                transform="translate(-7)"
                width="4"
                x="992"
                y="5"
              />
              <rect
                class="handle handle--n"
                cursor="ns-resize"
                height="6"
                rx="1"
                style=""
                transform="translate(-4)"
                width="10"
                x="989"
                y="2"
              />
              <rect
                class="handle handle--s"
                cursor="ns-resize"
                height="6"
                rx="1"
                style=""
                transform="translate(-4)"
                width="10"
                x="989"
                y="973"
              />
            </g>
          </g>
        </svg>
        <div
          class="c7"
        >
          <div
            class="c8"
            data-testid="zoom-out-button"
            disabled=""
          >
            <svg />
          </div>
          <span
            class="c9"
            data-testid="tooltip-container"
          >
            <div
              class="c10"
              data-testid="zoom-in-button"
            >
              <svg />
            </div>
          </span>
        </div>
      </div>
      <span
        class="c9"
        data-testid="tooltip-container"
      />
    </div>
  </div>
</body>
`;
